### ISSUE TITLE
About backward function of MaskedConv2d

### ISSUE THREAD

**Issue by ImCharlesY** [csmliu/AdaDSR#1]
Hi!

Thanks for sharing your impressive work!

I notice that currently you provide the sparse convolution implementation modified from [open-mmlab/mmdetection](https://github.com/open-mmlab/mmdetection/tree/master/mmdet/ops/masked_conv), which does not implement the backward function according to [masked_conv.py#L61](https://github.com/csmliu/AdaDSR/blob/master/masked_conv2d/masked_conv2d/masked_conv.py#L61). As a result, the gradient flows from output feature maps to both the input feature maps and the predicted masks are blocked.

Therefore, I am wondering how you make it possible to jointly train the proposed AdaDSR in an end-to-end manner under the supervision of the _reconstruction loss_ (Equation 6 in the paper)? I know that the adaptor module _P_ (notation from paper) can be trained by the _network depth loss_ (Equation 7). But it seems that the pre-trained model (except for the reconstruction module) will never be updated during the training procedure. Besides, since the masking operation in MaskedConv2d is an non-differentiable operation (it 'indexes' some rows in the data_col matrix), there is no gradient flow from the SR backbone to the adaptor module _P_. So it seems that the adaptor module is unaware of the 'SR image content' but just tries to meet the constraint of the desired network depth.

Thanks!

**Comment by csmliu** [comment]
Hi Charles,

Thanks for paying attention to our work.

The sparse_conv shares the same data structure with normal convolution layers, we can train the model with the standard nn.Conv2d provided by PyTorch itself, as I show in scripts/train_adarcan.sh. And during the inference phase, we can choose whether to use sparse_conv or not.

Therefore, both the adapter and the backbone model are updated during the training procedure.
We also tried to update the adapter only, but the results are much worse than the current version.

**Comment by csmliu** [comment]
I'm going to close this issue, if you have any further questions, please feel free to reopen it or open another issue.

**Comment by ImCharlesY** [comment]
Hi Liu,

Thanks for your quick response!

I recheck your codes, and find under the different configurations about `sparse_conv` option in [scripts/train_adarcan.sh](https://github.com/csmliu/AdaDSR/blob/master/scripts/train_adarcan.sh) and [scripts/test_adarcan.sh](https://github.com/csmliu/AdaDSR/blob/master/scripts/test_adarcan.sh#L25) AdaDSR is trained with standard convolution but tested with sparse convolution.

I also find that the predicted masks are multiplied with the output feature maps of the convolution according to [models/network.py#L483](https://github.com/csmliu/AdaDSR/blob/master/models/networks.py#L483). So the gradient flow can propogate from the backbone model to the adaptor as the above mechanism acts like some kind of a 'hard' spatial attention.

Besides, I am wondering if the pre-trained weights are key for the converagence or the performance of the model since it is worrying that the 'hard' masking operation might make the training procedure unstable. Have you ever tried training the model from scratch?

Thanks a lot!

**Comment by csmliu** [comment]
Yes, we train the models using standard convolution, and test via sparse convolution. You can test with standard convolution by removing `--sparse_conv True` or setting `--sparse_conv False`. Using standard convolution may be faster than sparse convolution, since the standard convolution is highly optimized and takes advantage of Winograd algorithm.
On the other hand, the sparse convolution does reduce some computations comparing to the im2col based standard convolution, but if you use standard convolution based sparse convolution, it brings some more computations.

It is crucial to multiply the predicted mask and the feature for the gradient flow. And the pre-trained weights are very important for the stable training. In our experiments, if no pre-trained model is provided, the model is very very hard to train and fails in most cases. However, with the pre-trained model, generally we can train the model stably.
